### PR TITLE
UI for User Roles

### PR DESCRIPTION
**Pull request by njirap**
Hi @mozbhearsum 

This PR is still under construction. I have created it so that the reviewing can be on a rolling basis and not request for a PR review on a gigantic lumpsum.

Each commit will be the ui for one of the CRUD operations on the User roles.

This first commit is for Viewing a User's roles

### CODE REVIEW

File: ui/app/js/controllers/user_edit_controller.js [hunk 1]
```diff
@@ -31,6 +31,15 @@ function ($scope, $modalInstance, CSRF, Permissions, user, users) {
     $scope.loading = false;
   });
 
+  $scope.user.roles = [];
+  Permissions.getUserRoles(user.username)
+  .success(function(response) {
+    $scope.user.roles = response.roles;
+  })
+  .error(function(response) {
+    console.error(response);
```
Comment (mozbhearsum) [inline ui/app/js/controllers/user_edit_controller.js]: We're trying to get away from silent errors - better to show an error in a modal dialog for this.

File: auslib/db.py [hunk 2]
```diff
@@ -1970,8 +1970,8 @@ def getOptions(self, username, permission, transaction=None):
             raise ValueError('Permission "%s" doesn\'t exist' % permission)
 
     def getUserRoles(self, username, transaction=None):
-        res = self.user_roles.select(where=[self.user_roles.username == username], columns=[self.user_roles.role], distinct=True, transaction=transaction)
-        return [r["role"] for r in res]
+        res = self.user_roles.select(where=[self.user_roles.username == username], columns=[self.user_roles.role, self.user_roles.data_version], distinct=True, transaction=transaction)
+        return [{"role": r["role"], "data_version": r["data_version"]} for r in res]
```
Comment (njirap) [inline auslib/db.py]: hi @mozbhearsum I require the roles to return the `data_version` because this is required when revoking a role from the ui

Comment (mozbhearsum) [inline auslib/db.py]: Good catch!

**Comment by njirap** [comment]
@mozbhearsum the latest commit ends my poking in the backend(hopefully)

File: auslib/test/admin/views/test_permissions.py [hunk 3]
```diff
@@ -518,8 +518,15 @@ class TestUserRolesAPI_JSON(ViewTest):
     def testGetRoles(self):
         ret = self._get("/users/bill/roles")
         self.assertStatusCode(ret, 200)
-        got = set(json.loads(ret.data)["roles"])
-        self.assertEquals(got, set(["releng", "qa"]))
+        got = json.loads(ret.data)["roles"]
+        self.assertEquals(got, [{"role": "qa", "data_version": 1},
+                          {"role": "releng", "data_version": 1}])
+
+    def testGettAllRoles(self):
```
Comment (mozbhearsum) [inline auslib/test/admin/views/test_permissions.py]: Nit: typo here - "Gett" vs "Get".

File: auslib/test/test_db.py [hunk 4]
```diff
@@ -3256,7 +3256,7 @@ def testHasPermissionNotAllowedByProduct(self):
 
     def testGetUserRoles(self):
         got = self.permissions.getUserRoles("bob")
-        self.assertEquals(set(got), set(["releng", "dev"]))
+        self.assertEquals(got, [{'data_version': 1, 'role': u'dev'}, {'data_version': 1, 'role': u'releng'}])
```
Comment (mozbhearsum) [inline auslib/test/test_db.py]: Please continue using a set here - the order in which these are returned from the database is not guaranteed.

File: auslib/admin/views/permissions.py [hunk 5]
```diff
@@ -180,6 +180,14 @@ def get(self, username):
         return jsonify({"roles": roles})
 
 
+class AllRolesView(AdminView):
+    """/users/roles"""
+
+    def get(self):
+        roles_list = dbo.permissions.getAllRoles()
+        return jsonify({"roles_list": roles_list})
```
Comment (mozbhearsum) [inline auslib/admin/views/permissions.py]: Nit: I think just plain "roles" would be a better name here - the type of the value is obvious to consumers.

File: ui/app/js/controllers/user_edit_controller.js (hunk 1 above)
Comment (mozbhearsum) [inline ui/app/js/controllers/user_edit_controller.js]: Don't forget to address this, please!

File: auslib/test/test_db.py (hunk 4 above)
Comment (njirap) [inline auslib/test/test_db.py]: I used `sorted` instead because a dictionary is not hashable

**Comment by njirap** [comment]
@mozbhearsum this PR is no longer a work in progress..

File: auslib/test/test_db.py (hunk 4 above)
Comment (mozbhearsum) [inline auslib/test/test_db.py]: 👍

File: ui/app/js/controllers/user_edit_controller.js [hunk 6]
```diff
@@ -31,12 +31,96 @@ function ($scope, $modalInstance, CSRF, Permissions, user, users) {
...
     value.options = value.options_as_json;
   });
 
+  $scope.grantRole = function() {
+    $scope.saving = true;
+    CSRF.getToken()
+    .then(function(csrf_token) {
+      Permissions.grantRole($scope.user.username, $scope.role.role, $scope.role.data_version, csrf_token)
+      .success(function(response) {
+        $scope.user.roles.push($scope.role);
```
Comment (mozbhearsum) [inline ui/app/js/controllers/user_edit_controller.js]: I think you need to parse the response for "new_data_version" and set it in the right place. Right now, if I try to revoke a role immediately after granting it, I get "data version is not a valid integer value".

Comment (mozbhearsum) [inline ui/app/js/controllers/user_edit_controller.js]: You should also update $scope.roles_list if the new Role isn't already in it.

**Comment by njirap** [comment]
@mozbhearsum changes made..

**Comment by mozbhearsum** [comment]
Looks great now!

**Pull request merged**